benchmarks/benchmarks/bench_function_base.py

```diff
@@ -96,46 +96,152 @@ def time_select_larger(self):
+        of swapped pairs is X% of the size of the array
+    random_unsorted_area_X_percent
+        This kind of array has random unsorted areas which
+        take up X% of the original array, the size of one
+        area is AREA_SIZE
+    random_bubble_X_fold
+        ?
+    """
     params = [
+        # In NumPy 1.17 and newer, 'merge' can be one of several
+        # stable sorts
         ['quick', 'merge', 'heap'],
-        ['float32', 'int32', 'uint32']
+        ['float64', 'int64', 'uint64'],
+        [
+            'ordered',
+            'reversed',
+            'uniform',
+            'sorted_block_10',
+            'sorted_block_100',
+            'sorted_block_1000',
+            'swapped_pair_1_percent',
+            'swapped_pair_10_percent',
+            'swapped_pair_50_percent',
+            'random_unsorted_area_50_percent',
+            'random_unsorted_area_10_percent',
+            'random_unsorted_area_1_percent',
+            'random_bubble_1_fold',
+            'random_bubble_5_fold',
+            'random_bubble_10_fold',
+        ],
     ]
-    param_names = ['kind', 'dtype']
-
-    def setup(self, kind, dtype):
-        self.e = np.arange(10000, dtype=dtype)
-        self.o = np.arange(10001, dtype=dtype)
-        np.random.seed(25)
-        np.random.shuffle(self.o)
-        # quicksort implementations can have issues with equal elements
-        self.equal = np.ones(10000, dtype=dtype)
-        self.many_equal = np.sort(np.arange(10000) % 10).astype(dtype)
-
-        try:
-            np.sort(self.e, kind=kind)
-        except TypeError:
-            raise NotImplementedError()
-
-    def time_sort(self, kind, dtype):
-        np.sort(self.e, kind=kind)
-
-    def time_sort_random(self, kind, dtype):
-        np.sort(self.o, kind=kind)
-
-    def time_sort_inplace(self, kind, dtype):
-        self.e.sort(kind=kind)
-
-    def time_sort_equal(self, kind, dtype):
-        self.equal.sort(kind=kind)
-
-    def time_sort_many_equal(self, kind, dtype):
-        self.many_equal.sort(kind=kind)
-
-    def time_argsort(self, kind, dtype):
-        self.e.argsort(kind=kind)
-
-    def time_argsort_random(self, kind, dtype):
-        self.o.argsort(kind=kind)
+    param_names = ['kind', 'dtype', 'array_type']
+
+    # The size of the benchmarked arrays.
+    ARRAY_SIZE = 10000
+    # The size of the unsorted area in the "random unsorted area"
+    # benchmarks
+    AREA_SIZE = 10
+    # ?
+    BUBBLE_SIZE = 10
+
+    def setup(self, kind, dtype, array_type):
+        np.random.seed(1234)
+        self.arr = getattr(self, 'array_' + array_type)(self.ARRAY_SIZE, dtype)
+
+    def time_sort_inplace(self, kind, dtype, array_type):
+        self.arr.sort(kind=kind)
+
+    def time_sort(self, kind, dtype, array_type):
+        np.sort(self.arr, kind=kind)
+
+    def time_argsort(self, kind, dtype, array_type):
+        np.argsort(self.arr, kind=kind)
+
+    def array_random(self, size, dtype):
```

Comment:
It _might_ make it easier to visually separate the tests and these generators by pulling the generator functions into the base class, or even using composition instead of inheritance:
```python
class SortGenerator:
     # all of the array_* functions

class Sort(Benchmark):
    def setup():
        self.arr = getattr(SortGenerator(), 'array_' + array_type)(self.ARRAY_SIZE, dtype)

Reply:
I did this. In addition I wrote a `__getattr__` that parses the array type for parametrized array types.